# Explore Any PulseLM Dataset

['afppgecg',
'bcg',
'bidmc',
'dalia',
'earset',
'mimicperform',
'ppgarrhythmia',
'ppgbp',
'sdb',
'sensors',
'uci',
'uqvitalsigns',
'utsappg',
'vitaldb',
'wesad',
'wildppg']


In [1]:
import os
import sys

import pandas as pd

from tqdm import tqdm

# Get the current working directory of the notebook
notebook_dir = os.getcwd()

# Add the parent directory to the system path
sys.path.append(os.path.join(notebook_dir, '../custom_utils'))

from data_processing import DataProcessing

/Users/detraviousjamaribrinkley/Documents/Development/research_labs/roeh_labs/health/closed_loop_wearable_therapies/PulseLM/.venv/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [2]:
pd.set_option('max_colwidth', 800)
# pd.set_option('display.max_columns', None)
# pd.set_option('display.max_rows', None)

## Load Data

In [3]:
# Load single dataset
# wesad_df = DataProcessing.load_pulselm_datasets(notebook_dir, 'wesad')

# Load specific datasets
# datasets = DataProcessing.load_pulselm_datasets(notebook_dir, ['wesad', 'dalia'])

# Load all datasets
all_datasets = DataProcessing.load_pulselm_datasets(notebook_dir)

Loaded afppgecg: 64725 samples
Loaded bcg: 521 samples
Loaded bidmc: 9412 samples
Loaded dalia: 29294 samples
Loaded earset: 1368 samples
Loaded mimicperform: 3239 samples
Loaded ppgarrhythmia: 36820 samples
Loaded ppgbp: 294 samples
Loaded sdb: 205618 samples
Loaded sensors: 1631 samples
Loaded uci: 89054 samples
Loaded uqvitalsigns: 25202 samples
Loaded utsappg: 12740 samples
Loaded vitaldb: 131324 samples
Loaded wesad: 2104 samples
Loaded wildppg: 180000 samples


## Note the Time Series Datasets

| Dataset | # Samples | Description (features) | Raw Time Series |
|---|---:|---|---|
| `afppgecg` | 64,725 | Features: `signal`, `text`, `qa`. PulseLM-formatted PPG/ECG samples paired with language description and QA labels; stored as task-ready segments rather than raw continuous recordings. | ❌ |
| `bcg` | 521 | Features: `signal`, `text`, `qa`. PulseLM-formatted BCG samples with text and QA annotations; not raw continuous time series in this version. | ❌ |
| `bidmc` | 9,412 | Features: `signal`, `text`, `qa`. Physiological waveform samples with associated text and QA labels; provided as processed task-ready segments. | ❌ |
| `dalia` | 29,294 | Features: `signal`, `text`, `qa`. Wearable physiological sensor samples with descriptive text and QA labels; not raw continuous recordings in PulseLM format. | ❌ |
| `earset` | 1,368 | Features: `signal`, `text`, `qa`. Ear-based physiological sensor samples with text and QA annotations; appears task-formatted rather than raw. | ❌ |
| `mimicperform` | 3,239 | Features: `signal`, `text`, `qa`. Clinical physiological waveform/perfusion samples with text and QA metadata; exported here as processed segments. | ❌ |
| `ppgarrhythmia` | 36,820 | Features: `signal`, `text`, `qa`. PPG arrhythmia-related waveform samples with text and QA labels; segmented for benchmarking rather than raw continuous series. | ❌ |
| `ppgbp` | 294 | Features: `signal`, `text`, `qa`. PPG-based blood pressure samples with text descriptions and QA annotations; stored as signal windows, not raw continuous data. | ❌ |
| `sdb` | 205,618 | Features: `signal`, `text`, `qa`. Large physiological signal dataset with language and QA metadata; represented as task-ready segments. | ❌ |
| `sensors` | 1,631 | Features: `signal`, `text`, `qa`. General sensor-based physiological samples with text and QA labels; not raw continuous recordings in this exported form. | ❌ |
| `uci` | 89,054 | Features: `signal`, `text`, `qa`. UCI-derived physiological or wearable sensor signal samples with text and QA annotations; already segmented. | ❌ |
| `uqvitalsigns` | 25,202 | Features: `signal`, `text`, `qa`. Vital sign waveform/sensor samples paired with text descriptions and QA labels; task-formatted rather than raw continuous series. | ❌ |
| `utsappg` | 12,740 | Features: `signal`, `text`, `qa`. Smartphone/app-based PPG samples with text and QA annotations; segmented and preformatted for downstream tasks. | ❌ |
| `vitaldb` | 131,324 | Features: `signal`, `text`, `qa`. Clinical vital waveform samples from VitalDB with text and QA labels; not raw continuous time series in the PulseLM export. | ❌ |
| `wesad` | 2,104 | Features: `signal`, `text`, `qa`. Wearable stress/activity physiological samples with text description and QA labels; transformed into task-ready segments in this version. | ❌ |
| `wildppg` | 180,000 | Features: `signal`, `text`, `qa`. Large in-the-wild PPG signal samples with text and QA annotations; appears segmented/windowed for benchmarking use. | ❌ |

### Initial note
For our purposes, we want **raw continuous time series** suitable for resampling, filtering, DC-offset removal, windowing, and normalization. In the PulseLM-formatted datasets, **none of the exported signals are raw continuous time series**; they are already **segmented and task-formatted**, so they should not be used directly for the full preprocessing pipeline.

In [18]:
all_datasets['wildppg']

,signal,text,qa
0,"[0.47123631834983826, 0.4532877802848816, 0.4364531934261322, 0.41913726925849915, 0.4021887481212616, 0.3850189447402954, 0.3678913116455078, 0.350566565990448, 0.33325815200805664, 0.3161320388317108, 0.29963362216949463, 0.2843005657196045, 0.2707534730434418, 0.25959512591362, 0.2513088583946228, 0.24616600573062897, 0.24415990710258484, 0.24498705565929413, 0.24808043241500854, 0.25269147753715515, 0.2580026090145111, 0.2632559537887573, 0.2678830623626709, 0.27161189913749695, 0.2745237648487091, 0.27703315019607544, 0.2797880470752716, 0.2835162580013275, 0.28886446356773376, 0.29628023505210876, 0.30596867203712463, 0.31792813539505005, 0.3320411741733551, 0.34818363189697266, 0.36632445454597473, 0.3865507245063782, 0.40903332829475403, 0.43393123149871826, 0.4612658619880676,...",PPG signal recorded from wrist position. Heart rate is 88 bpm (normal). Mean RR interval is 686 ms. RMSSD is 10.8 ms indicating reduced parasympathetic activity. SDNN is 7.6 ms showing reduced heart rate variability. pNN50 is 0.0%.,"{""heart_rate_category"": {""question"": ""Provide the clinical heart rate category."", ""answer"": ""normal""}, ""hrv_sdnn_category"": {""question"": ""What is the HRV SDNN category for this segment?"", ""answer"": ""low""}, ""hrv_rmssd_category"": {""question"": ""What is the parasympathetic activity level (RMSSD)?"", ""answer"": ""low""}, ""hrv_pnn50_category"": {""question"": ""Provide the pNN50 category."", ""answer"": ""low""}}"
1,"[0.664634108543396, 0.6832813620567322, 0.698003351688385, 0.7118421792984009, 0.721674919128418, 0.7282105684280396, 0.730369508266449, 0.7288196682929993, 0.7236606478691101, 0.7156867980957031, 0.7054551243782043, 0.6935745477676392, 0.6804651618003845, 0.6663587689399719, 0.6513438820838928, 0.6354784369468689, 0.6189411282539368, 0.6021642684936523, 0.5858896374702454, 0.5711119174957275, 0.5589194893836975, 0.550280749797821, 0.5458381772041321, 0.5457647442817688, 0.5497045516967773, 0.5567948818206787, 0.5657501220703125, 0.5749879479408264, 0.5827852487564087, 0.587449848651886, 0.587493360042572, 0.5817911028862, 0.5697178840637207, 0.5512248873710632, 0.5268325209617615, 0.4976152181625366, 0.465079665184021, 0.4310094118118286, 0.3972788155078888, 0.36565327644348145, 0.337...",PPG signal recorded from wrist position. Heart rate is 88 bpm (normal). Mean RR interval is 682 ms. RMSSD is 16.2 ms indicating reduced parasympathetic activity. SDNN is 16.7 ms showing reduced heart rate variability. pNN50 is 0.0%.,"{""heart_rate_category"": {""question"": ""Is the heart rate normal, bradycardic, or tachycardic?"", ""answer"": ""normal""}, ""hrv_sdnn_category"": {""question"": ""How would you categorize SDNN for this PPG?"", ""answer"": ""low""}, ""hrv_rmssd_category"": {""question"": ""Determine the RMSSD classification."", ""answer"": ""low""}, ""hrv_pnn50_category"": {""question"": ""Is pNN50 low, normal, or high in this sample?"", ""answer"": ""low""}}"
2,"[0.4624752700328827, 0.42166846990585327, 0.38530436158180237, 0.35416096448898315, 0.3315952718257904, 0.3183230757713318, 0.31527072191238403, 0.3217315971851349, 0.3366573750972748, 0.3581443428993225, 0.3840903639793396, 0.41226691007614136, 0.4406493306159973, 0.4676310122013092, 0.4921777844429016, 0.5138919353485107, 0.5329833626747131, 0.5501578450202942, 0.5664457678794861, 0.5829997062683105, 0.6008891463279724, 0.6209186315536499, 0.6434926390647888, 0.6685466170310974, 0.6955534219741821, 0.7236043214797974, 0.75154709815979, 0.7781544327735901, 0.8022918701171875, 0.8230617046356201, 0.8399031162261963, 0.8526394963264465, 0.8614562153816223, 0.8668398261070251, 0.8694905042648315, 0.8701391816139221, 0.8694277405738831, 0.8678153157234192, 0.8655496835708618, 0.8627052903...",PPG signal recorded from wrist position. Heart rate is 87 bpm (normal). Mean RR interval is 694 ms. RMSSD is 6.9 ms indicating reduced parasympathetic activity. SDNN is 10.0 ms showing reduced heart rate variability.